### ISSUE TITLE
Uninitialized unique_ptr

### ISSUE THREAD

**Issue by shirolimit** [boost-ext/di#221]
Hi, Krzysztof!

Why I cannot create object with unique_ptr member, that is not initialized explicitly in ctor. 
It can be initialized later for example using setter.
Here is the code:

```
class Interface1 {};

class Interface2 {};

class Impl1: public Interface1 {};

class Impl2: public Interface2 {};

class App
{
public:
    explicit App(Interface1* if1)
        : _if1(if1)
    {
    }

    virtual ~App() noexcept = default;

private:
    Interface1* _if1;
    std::unique_ptr<Interface2> _if2;
};

namespace di = boost::di;

int main(int argc, char* argv[])
{

    auto injector = di::make_injector(
        di::bind<Interface1>.to<Impl1>()
    );
    auto house = injector.create<App*>();

    return 0;
}
```

Compiler output is as follows (GCC 5.2.1):

`error(_ = "type is not bound, did you forget to add: 'di::bind<interface>.to<implementation>()'?");`

If I change _if2 type to plain pointer or shared_ptr, the code compiles without errors.

**Comment by kris-jusiak** [comment]
Thanks @shirolimit for the example. Your code is obviously valid and there is an issue in DI. I will try to fix it ASAP. There is another issue related to similar problem here #219.

For now, there are a few workarounds which may temporary fix the problem until proper solution will not be released.
- The problem occurs only when constructor has one argument, so an easy fix would be to add more constructor parameters

``` cpp
class App
{
public:
    explicit App(Interface1* if1, ...) // due to ... di will be able to use this constructor instead of having an ambiguity with move/copy constructor
        : _if1(if1)
    {
    }

    virtual ~App() noexcept = default;

private:
    Interface1* _if1;
    std::unique_ptr<Interface2> _if2;
};
```
- If you can make App movable it will help too

``` cpp
class App
{
public:
    explicit App(Interface1* if1)
         : _if1(if1)
    {
    }

    App(App&&) = default;
    virtual ~App() noexcept = default;

private:
    Interface1* _if1;
    std::unique_ptr<Interface2> _if2;
};
```
- you can also say explicitly what parameters should be used by the constructor

``` cpp
class App
{
public:
    BOOST_DI_INJECT(explicit App, Interface1* if1)
         : _if1(if1)
    {
    }

    App(App&&) = default;
    virtual ~App() noexcept = default;

private:
    Interface1* _if1;
    std::unique_ptr<Interface2> _if2;
};
```

**Comment by shirolimit** [comment]
Wow. @krzysztof-jusiak thank you for detailed answer and for such a great library!

**Comment by kris-jusiak** [comment]
Cheers @shirolimit; Anyway it should be fixed now by https://github.com/boost-experimental/di/commit/a809130c23bb1996d1a9a8bd8c9fbc3e75e01a82 (tested on gcc-6 and clang-3.7). Currently travis is checking all sort of compilers. If you can confirm that the newest master actually fix your issue it would be awesome. Thanks.

After #219 will be fixed I will release a new version (1.0.2).
